### PR TITLE
Licensing and Software: add Licensing properties to SoftwareArtifact

### PR DESCRIPTION
**Pull request by swinslow**
Signed-off-by: Steve Winslow <steve@swinslow.net>

### CODE REVIEW

**Comment by swinslow** [comment]
cc @jlovejoy @kestewart @maxhbr @zvr @goneall @armintaenzertng 

I am not certain whether @zvr will be submitting a PR to make the Licensing properties into Relationships in time for the RC release. In particular, I am not sure whether this is currently feasible since License / LicenseAddition are not Elements (or, more to the point, since LicenseField is not an Element).

Given that, this is my best guess at how we might represent that the Licensing properties are _optional_ properties for a SoftwareArtifact. This may break the parser. I've reviewed the other classes in Core and Software, and I don't see a clear way to represent this, so I defer to @maxhbr and others if there's a different way that this should be specified.

Please feel free to revise as appropriate, but there should be some way to indicate that a SoftwareArtifact can use the Licensing properties -- if these won't be included as Relationships for the RC.

**Review by kestewart — APPROVED** [review]
This looks like a pragmatic way to handle this issue for now, and will unblock the AI & Dataset.

**Comment by zvr** [comment]
The PR will also be making Elements the corresponding classes from Licensing ;-).

File: model/Software/Classes/SoftwareArtifact.md
Comment (zvr) [inline model/Software/Classes/SoftwareArtifact.md:29]: The property is for this class, not an external one.
It should be `concludedLicense`, and have a description in /Software/Properties.

If it's not used anywhere else in the Licensing namespace, it can be deleted from there.

The same for all of them.

File: model/Software/Classes/SoftwareArtifact.md
Comment (zvr) [inline model/Software/Classes/SoftwareArtifact.md:30]: `/Licensing/LicenseField`, it's not under Core

**Review by zvr — CHANGES_REQUESTED** [review]

**Comment by kestewart** [comment]
@swinslow - see change request from Alexios.

File: model/Software/Classes/SoftwareArtifact.md
Comment (swinslow) [reply model/Software/Classes/SoftwareArtifact.md:29]: Sorry, just to clarify, does this mean that there wouldn't be any classes or properties defined in the Licensing profile? All of the classes are used to build up what is a License, so would all of them need to be in Software instead of Licensing?

**Comment by swinslow** [comment]
> The PR will also be making Elements the corresponding classes from Licensing ;-).

What are the other implications of this change? What does it mean to make Licenses into Elements?

**Comment by swinslow** [comment]
> @swinslow - see change request from Alexios.

@zvr @kestewart Thanks -- I'll update to change `/Core/Licensing` to just `/Licensing` throughout.

But, for moving e.g. `concludedLicense` from the `Licensing` profile into `Software`, I don't understand what that means. Does that mean moving all of the classes and properties from `Licensing` into `Software`, since they build on top of one another?

File: model/Software/Classes/SoftwareArtifact.md
Comment (swinslow) [reply model/Software/Classes/SoftwareArtifact.md:33]: It's still up to an SPDX data creator if they want to explicitly specify NOASSERTION, so I would say it should still be an option.

**Comment by swinslow** [comment]
@zvr @kestewart It looks to me like [Element](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/Element.md) contains a bunch of properties that would not necessarily apply to Licenses, and never have in the past.

I don't think we can make a License into an Element without a more detailed discussion and consensus for what that means.

I'm inclined to say that we should keep Licenses as they are, as properties, for this RC since that at least maintains parity with 2.3 and all prior SPDX specs.

File: model/Software/Classes/SoftwareArtifact.md
Comment (maxhbr) [reply model/Software/Classes/SoftwareArtifact.md:33]: But, could we make that a serialization problem? For the model we could say that `NoAssertion` and `null` are the same. And we could give in serializations the option to use `NoAssertion` in place of `null`

File: model/Software/Classes/SoftwareArtifact.md
Comment (maxhbr) [reply model/Software/Classes/SoftwareArtifact.md:29]: this is written within the scope of the Software profile. So you need to qualify values from Licensing. I think correct would be:
```suggestion
- /Licensing/concludedLicense
```

File: model/Software/Classes/SoftwareArtifact.md
Comment (maxhbr) [reply model/Software/Classes/SoftwareArtifact.md]: maybe not blocking, but I think this builds up dept, which we might not be able to get rid of, without breaking changes (which we should discuss before release)

Comment (maxhbr) [reply model/Software/Classes/SoftwareArtifact.md]: I created https://github.com/spdx/spdx-3-model/issues/289 for that

Comment (swinslow) [reply model/Software/Classes/SoftwareArtifact.md]: When it comes to concludedLicense or declaredLicense, there are two possibilities (leaving aside `NONE` and an actual license / license combination):
1. explicitly state `NOASSERTION` - this means the field is present and explicitly given as NOASSERTION; or
2. not specified - this means the field is _not_ present

For option 2, this was introduced as an option in SPDX 2.3 -- see e.g. https://spdx.github.io/spdx-spec/v2.3/package-information/#713-concluded-license-field: "If the Concluded License field is not present in a package, it implies an equivalent meaning to NOASSERTION."

This was added because people complained about having to have `LicenseConcluded: NOASSERTION` be listed for every package, file or snippet.

So I would still suggest allowing either serialization outcome. Either it is present as NOASSERTION, or it is absent and a consumer would interpret that silence as being semantically equivalent to NOASSERTION (if they care about it at all).

If that's something that we should add to the `concludedLicense` and `declaredLicense` properties, I can submit a PR to add that.

**Comment by maxhbr** [comment]
Maybe off-topic in this thread:

> @zvr @kestewart It looks to me like [Element](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/Element.md) contains a bunch of properties that would not necessarily apply to Licenses, and never have in the past.

I think that there is a serialization issue right now. The current approach for serialization is, that we "just" have a list of Elements. This means, there is no place for License objects. If they would derive from Element this would resolve the issue.

File: model/Software/Classes/SoftwareArtifact.md
Comment (swinslow) [reply model/Software/Classes/SoftwareArtifact.md]: Thanks @maxhbr -- I think I've updated it correctly to handle this, grateful if you can let me know if it still looks incorrect after the latest commit.

**Comment by kestewart** [comment]
Changes that @zvr requested have been made.   
@maxhbr please open an issue so that we can track having this discussion and resolve it before the 3.0 is marked as final.   I don't think it blocks us getting the model out, as it's a searialization issue.

**Pull request merged**
